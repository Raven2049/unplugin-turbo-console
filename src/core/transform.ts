import { basename, extname } from 'node:path'
import { ScriptTarget, createSourceFile, forEachChild, isCallExpression } from 'typescript'
import MagicString from 'magic-string'
import { parse as vueParser } from 'vue/compiler-sfc'
import type { Options, TraverseAST } from '../types'
import { getConsoleStyle } from './utils'

function traverseAST(args: TraverseAST) {
  const { node, s, id, options, offset = 0 } = args
  const fileName = basename(id)
  const fileType = extname(id)

  if (isCallExpression(node) as any) {
    if (node?.expression?.expression?.escapedText === 'console' && node?.expression?.name?.escapedText === 'log') {
      const argumentStart = node.arguments[0].pos + offset
      const argumentEnd = node.arguments[node.arguments.length - 1].end + offset
      const lineNumber = s.slice(0, argumentStart)?.split('\n').length

      const argsName = s.slice(argumentStart, argumentEnd).toString()
        .replaceAll('`', '')
        .replaceAll('\n', '')
        .replaceAll('\"', '')

      let { suffix, prefix } = options || {}
      suffix = suffix ? `\\n ${suffix}` : ''
      prefix = prefix ? `${prefix} \\n` : ''

      s.appendLeft(argumentStart, `"${prefix} %c${fileName}:${lineNumber} ~ ${argsName}","${getConsoleStyle(fileType)}",`)
      s.appendRight(argumentEnd, `,"${suffix}"`)
    }
  }

  forEachChild(node, childNode => traverseAST({
    node: childNode,
    s,
    id,
    options,
    offset,
  }))
}

export function transformer(code: string, id: string, options: Options) {
  if (['.js', '.ts', '.tsx', '.jsx'].includes(extname(id))) {
    const ast = createSourceFile(id, code, ScriptTarget.Latest)
    const s = new MagicString(code)
    traverseAST({
      node: ast,
      s,
      id,
      options,
    })
    return s.toString()
  }
  else if (extname(id) === '.vue') {
    const { descriptor } = vueParser(code, {
      sourceMap: true,
    })
    const s = new MagicString(code)
    if (descriptor.scriptSetup?.content) {
      const ast = createSourceFile(id, descriptor.scriptSetup?.content, ScriptTarget.Latest)
      traverseAST({
        node: ast,
        s,
        id,
        options,
        offset: descriptor.scriptSetup.loc.start.offset,
      })
    }

    if (descriptor.script?.content) {
      const ast = createSourceFile(id, descriptor.script?.content, ScriptTarget.Latest)
      traverseAST({
        node: ast,
        s,
        id,
        options,
        offset: descriptor.script.loc.start.offset,
      })
    }

    return s.toString()
  }
  return code
}
