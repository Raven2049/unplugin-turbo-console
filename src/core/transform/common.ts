import { cwd } from 'node:process'
import { Buffer } from 'node:buffer'
import { basename, extname, relative } from 'pathe'
import type { Node } from '@babel/types'
import type { GenContext } from '../../types'
import { getConsoleStyle, launchEditorStyle } from '../utils'

export function genConsoleString(genContext: GenContext) {
  const { options, originalColumn, originalLine, argType, id } = genContext
  let { argsName } = genContext
  const { prefix, suffix, disableLaunchEditor, port, disableHighlight } = options
  const _prefix = prefix ? `console.log("${prefix}");` : ''
  const consoleEndString = suffix ? `;console.groupEnd();console.log("${suffix}")})()` : ';console.groupEnd()})()'

  const urlObject = new URL(id, 'file://')
  const filePath = urlObject.pathname
  const fileName = basename(filePath)
  const fileType = extname(filePath)

  if (argsName?.length > 30)
    argsName = `${argsName.slice(0, 30)}...`

  // not output when argtype is string or number
  const lineInfo = `%cðŸš€ ${fileName}:${originalLine}${['StringLiteral', 'NumericLiteral'].includes(argType) ? '' : ` ~ ${argsName}`}`
  const codePosition = `${relative(cwd(), filePath)}:${originalLine}:${(originalColumn || 0) + 1}`

  const launchEditorString = `%cðŸ”¦ http://localhost:${port}/client#${Buffer.from(codePosition, 'utf-8').toString('base64')}`
  let consoleStartString = ''

  if (!disableHighlight && !disableHighlight)
    consoleStartString = `-(()=>{${_prefix}console.group("${lineInfo}${launchEditorString}","${getConsoleStyle(fileType)}","${launchEditorStyle}");`

  else if (disableHighlight && !disableLaunchEditor)
    consoleStartString = `-(()=>{${_prefix}console.group("${launchEditorString}","${launchEditorStyle}");`

  else if (!disableHighlight && disableLaunchEditor)
    consoleStartString = `-(()=>{${_prefix}console.group("${lineInfo}","${getConsoleStyle(fileType)}");`
  else
    consoleStartString = `-(()=>{${_prefix}console.group();`

  return {
    consoleStartString,
    consoleEndString,
  }
}

export function isConsoleExpression(node: Node) {
  return node.type === 'CallExpression'
    && node.callee.type === 'MemberExpression'
    && node.callee.object.type === 'Identifier'
    && node.callee.object.name === 'console'
    && node.callee.property.type === 'Identifier'
    && node.callee.property.name === 'log'
    && node.arguments?.length > 0
}
