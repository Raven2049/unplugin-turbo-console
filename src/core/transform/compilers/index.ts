import type { Compiler, CompileResult, Context } from '../../../types'
import { svelteCompiler } from './svelte'
import { vanillaCompiler } from './vanilla'
import { vue2Compiler } from './vue2'
import { vue3Compiler } from './vue3'

export const compilers: Record<Compiler, (context: Context) => Promise<CompileResult>> = {
  vue3: vue3Compiler,
  vue2: vue2Compiler,
  vanilla: vanillaCompiler,
  svelte: svelteCompiler,
}
