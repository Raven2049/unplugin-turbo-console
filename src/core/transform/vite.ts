import MagicString from 'magic-string'
import type { WithScope } from 'ast-kit'
import { babelParse, getLang, walkAST } from 'ast-kit'
import type { Node } from '@babel/types'
import { SourceMapConsumer } from 'source-map-js'
import type { Context } from '../../types'
import { genConsoleString, isConsoleExpress } from './common'

export function viteTransform(context: Context) {
  const { code, id, pluginContext, options } = context
  const magicString = new MagicString(code)
  const program = babelParse(code, getLang(id), {
    sourceFilename: id,
  })

  walkAST<WithScope<Node>>(program, {
    enter(node) {
      if (isConsoleExpress(node)) {
        const { line, column } = node.loc!.start
        // @ts-expect-error any
        const args = node.arguments
        // @ts-expect-error any
        const sourceMap = pluginContext.getCombinedSourcemap()

        const consumer = new SourceMapConsumer(sourceMap)

        const { line: originalLine, column: originalColumn } = consumer.originalPositionFor({
          line,
          column,
        })

        const argsStart = args[0].start!
        const argsEnd = args[args.length - 1].end!
        const argType = args[0].type

        const argsName = magicString.slice(argsStart, argsEnd)
          .toString()
          .replace(/`/g, '')
          .replace(/\n/g, '')
          .replace(/"/g, '')

        const { consoleString, _suffix } = genConsoleString({
          options,
          originalLine,
          originalColumn,
          argType,
          argsName,
          id,
        })

        magicString
          .appendLeft(argsStart, consoleString)
          .appendRight(argsEnd, `,"${_suffix}"`)
      }
    },
  })

  return {
    code: magicString.toString(),
    map: magicString.generateMap({ source: id }),
  }
}
