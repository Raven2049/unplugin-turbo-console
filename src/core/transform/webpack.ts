import { basename, extname } from 'node:path'
import { parse } from '@vue/compiler-sfc'
import type { WithScope } from 'ast-kit'
import { babelParse, getLang, walkAST } from 'ast-kit'
import type { Node } from '@babel/types'
import MagicString from 'magic-string'
import type { Context } from '../../types'
import { genConsoleString, isConsoleExpress } from './common'

const vuePatterns = [/\.vue$/, /\.vue\?vue/, /\.vue\?v=/]

export function webpackTransform(context: Context) {
  const { id, code, options } = context
  let scriptString = code
  let scriptLang = getLang(context.id)
  let vueSfcLocStart = {
    line: 0,
    column: 0,
    offset: 0,
  }
  const magicString = new MagicString(code)
  if (vuePatterns.some(pattern => pattern.test(id))) {
    const { descriptor, errors } = parse(code, {
      filename: id,
    })

    if (errors.length === 0) {
      if (descriptor.script) {
        scriptString = descriptor.script.content
        scriptLang = descriptor.script.lang || ''
        vueSfcLocStart = descriptor.script.loc.start
      }

      else if (descriptor.scriptSetup) {
        scriptString = descriptor.scriptSetup.content
        scriptLang = descriptor.scriptSetup.lang || ''
        vueSfcLocStart = descriptor.scriptSetup.loc.start
      }

      vueSfcLocStart.line--
    }
  }

  const program = babelParse(scriptString, scriptLang, {
    sourceFilename: id,
  })

  walkAST<WithScope<Node>>(program, {
    enter(node) {
      if (isConsoleExpress(node)) {
        const urlObject = new URL(id, 'file://')
        const fileName = basename(urlObject.pathname)
        const fileType = extname(urlObject.pathname)

        const { line, column } = node.loc!.start
        // @ts-expect-error any
        const args = node.arguments

        const argsStart = args[0].start! + vueSfcLocStart.offset
        const argsEnd = args[args.length - 1].end! + vueSfcLocStart.offset
        const argType = args[0].type

        const argsName = magicString.slice(argsStart, argsEnd)
          .toString()
          .replace(/`/g, '')
          .replace(/\n/g, '')
          .replace(/"/g, '')

        const originalLine = line + vueSfcLocStart.line
        const originalColumn = column

        const { consoleString, _suffix } = genConsoleString({
          options,
          fileName,
          originalLine,
          originalColumn,
          argType,
          filePath: urlObject.pathname,
          argsName,
          fileType,
        })

        magicString
          .appendLeft(argsStart, consoleString)
          .appendRight(argsEnd, `,"${_suffix}"`)
      }
    },
  })

  return {
    code: magicString.toString(),
    map: magicString.generateMap({ source: id }),
  }
}
