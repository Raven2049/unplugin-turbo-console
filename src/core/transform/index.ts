import type { Context } from './../../types'
import MagicString from 'magic-string'
import { parseSync } from 'oxc-parser'
import { genConsoleString, getCompiler, getLineAndColumn, isConsoleExpression, isPluginDisable } from '../utils'
import { walk } from '../utils/walker'
import { compilers } from './compilers'

export async function transform(context: Context) {
  const { code, id, options } = context
  const magicString = new MagicString(code)

  const compiler = await getCompiler(id)

  if (!compiler) {
    return {
      code: magicString.toString(),
      map: magicString.generateMap({
        source: id,
        file: id,
        includeContent: true,
        hires: true,
      }),
    }
  }

  const compileResult = await compilers[compiler](context)

  const ast = parseSync(id, compileResult.script, {
    lang: (compileResult.lang || 'js') as 'js' | 'jsx' | 'ts' | 'tsx' | undefined,
    sourceType: 'module',
  })

  if (isPluginDisable({
    comments: ast.comments || [],
    originalLine: 1,
    script: compileResult.script,
    id,
    type: 'top-file',
    compiler,
  })) {
    return {
      code: magicString.toString(),
      map: magicString.generateMap({
        source: id,
        file: id,
        includeContent: true,
        hires: true,
      }),
    }
  }

  walk(ast.program, {
    enter(node) {
      if (isConsoleExpression(node)) {
        const expressionStart = node.start
        const expressionEnd = node.end

        const originalExpression = magicString.slice(expressionStart, expressionEnd)

        if (originalExpression.includes('%c'))
          return false

        const { line, column } = getLineAndColumn(compileResult.script, expressionStart)

        const originalLine = line + compileResult.line
        const originalColumn = column

        if (isPluginDisable({
          comments: ast.comments || [],
          script: compileResult.script,
          originalLine: line,
          id,
          type: 'inline-file',
          compiler,
        })) {
          return false
        }

        // @ts-expect-error any
        const args = node.arguments

        const argsStart = args[0].start! + compileResult.offset
        const argsEnd = args[args.length - 1].end! + compileResult.offset
        const argType = args[0].type

        const argsName = magicString.slice(argsStart, argsEnd)
          .toString()
          .replace(/`/g, '')
          .replace(/\n/g, '')
          .replace(/"/g, '')

        const { consoleString, _suffix } = genConsoleString({
          options,
          originalLine,
          originalColumn,
          argType,
          argsName,
          id,
        })

        consoleString && magicString.appendLeft(argsStart, consoleString)
        _suffix && magicString.appendRight(argsEnd, `,"${_suffix}"`)
      }
    },
  })

  return {
    code: magicString.toString(),
    map: magicString.generateMap({
      source: id,
      file: id,
      includeContent: true,
      hires: true,
    }),
  }
}
