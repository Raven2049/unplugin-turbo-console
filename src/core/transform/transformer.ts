import type { Context } from '../../types'
import { viteTransform } from './vite'
import { webpackTransform } from './webpack'

export function transformer(context: Context) {
  const { code, meta } = context
  if (meta!.framework === 'webpack') {
    // only works on webpack development mode
    if (meta!.webpack.compiler.options.mode === 'development')
      return webpackTransform(context)
    return code
  }
  else if (meta!.framework === 'rspack') {
    if (meta!.rspack.compiler.options.mode === 'development')
      return webpackTransform(context)
    return code
  }
  else {
    return viteTransform(context)
  }
}
