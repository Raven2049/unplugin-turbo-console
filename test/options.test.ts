import { join } from 'pathe'
import { describe, expect, it, vi } from 'vitest'
import { webpackTransform } from '../src/core/transform/webpack'
import { optionWithDisableLaunchEditor, optionWithPrefix } from './fixtures/option'

vi.mock('node:process', () => {
  return {
    cwd: vi.fn(() => join('/', 'mock', 'path')),
  }
})

describe('options', () => {
  it('perfix suffix', () => {
    expect(
      webpackTransform(
        optionWithPrefix,
      ),
    ).toMatchSnapshot()
  })

  it('disable launch editor', () => {
    expect(
      webpackTransform(
        optionWithDisableLaunchEditor,
      ),
    ).toMatchSnapshot()
  })
})
