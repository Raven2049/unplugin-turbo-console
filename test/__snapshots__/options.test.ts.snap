// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`options > disable all 1`] = `
{
  "code": "-(()=>{console.group();console.log('hello javascript');console.groupEnd()})()",
  "map": SourceMap {
    "file": undefined,
    "mappings": "uBAAA",
    "names": [],
    "sources": [
      "/home/runner/work/unplugin-turbo-console/src/App.vue",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`options > disable highlight editor 1`] = `
{
  "code": "-(()=>{console.group("%cðŸ”¦ http://localhost:undefined/client#Li4vLi4vaG9tZS9ydW5uZXIvd29yay91bnBsdWdpbi10dXJiby1jb25zb2xlL3NyYy9BcHAudnVlOjE6MQ==","background: #00DC8250;padding:2px 5px;border-radius:0 3px 3px 0;margin-bottom:5px");console.log('hello javascript');console.groupEnd()})()",
  "map": SourceMap {
    "file": undefined,
    "mappings": "wOAAA",
    "names": [],
    "sources": [
      "/home/runner/work/unplugin-turbo-console/src/App.vue",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`options > disable launch editor 1`] = `
{
  "code": "-(()=>{console.group("%cðŸš€ App.vue:1%cðŸ”¦ http://localhost:undefined/client#Li4vLi4vaG9tZS9ydW5uZXIvd29yay91bnBsdWdpbi10dXJiby1jb25zb2xlL3NyYy9BcHAudnVlOjE6MQ==","padding:2px 5px; border-radius:3px 0 0 3px;margin-bottom:5px;color: #fff; background: #4FC08D","background: #00DC8250;padding:2px 5px;border-radius:0 3px 3px 0;margin-bottom:5px");console.log('hello javascript');console.groupEnd()})()",
  "map": SourceMap {
    "file": undefined,
    "mappings": "sVAAA",
    "names": [],
    "sources": [
      "/home/runner/work/unplugin-turbo-console/src/App.vue",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`options > empty option 1`] = `
{
  "code": "-(()=>{console.group("%cðŸš€ App.vue:1%cðŸ”¦ http://localhost:undefined/client#Li4vLi4vaG9tZS9ydW5uZXIvd29yay91bnBsdWdpbi10dXJiby1jb25zb2xlL3NyYy9BcHAudnVlOjE6MQ==","padding:2px 5px; border-radius:3px 0 0 3px;margin-bottom:5px;color: #fff; background: #4FC08D","background: #00DC8250;padding:2px 5px;border-radius:0 3px 3px 0;margin-bottom:5px");console.log('hello javascript');console.groupEnd()})()",
  "map": SourceMap {
    "file": undefined,
    "mappings": "sVAAA",
    "names": [],
    "sources": [
      "/home/runner/work/unplugin-turbo-console/src/App.vue",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`options > perfix suffix 1`] = `
{
  "code": "-(()=>{console.log("xxxx");console.group();console.log('hello javascript');console.groupEnd();console.log("yyy")})()",
  "map": SourceMap {
    "file": undefined,
    "mappings": "2CAAA",
    "names": [],
    "sources": [
      "/home/runner/work/unplugin-turbo-console/src/App.vue",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;
